Sources/PackageLoading/TargetSourcesBuilder.swift

```diff
@@ -86,13 +86,17 @@ public struct TargetSourcesBuilder {
         
         self.excludedPaths.forEach { exclude in
             if let message = validTargetPath(at: exclude) {
-                self.diags.emit(warning: "Invalid Exclude: \(exclude): \(message)")
+                if !self.diags.diagnostics.contains(where: { $0.localizedDescription == "Invalid Exclude: \(message) '\(exclude)'" }) {
+                    self.diags.emit(warning: "Invalid Exclude: \(message) '\(exclude)'")
+                }
```

Comment:
probably more robust to construct the message once and then use that variable in both places, e.g.

```
let message = "Invalid Exclude: \(message) '\(exclude)'"
if !self.diags.diagnostics.contains(where: { $0.localizedDescription == message }) {
   self.diags.emit(warning: message)
}
```

Reply:
the other question here is mostly for @abertelrud which knows the `DiagnosticsEngine`: is there a more efficient way to do this? scanning the `self.diags.diagnostics` would be pretty bad from a performance point of view. One could also make the case that `self.diags.diagnostics` should be an ordered set rather than an array which could avoid these type of issues in the first place, not sure if this was considered in the past

Reply:
I agree that there should be something more efficient here, though it's usually somewhat of an antipattern for logic to ask the diagnostics engine whether it has emitted a particular diagnostic already — usually it would keep track of that itself.

In this case, could TargetSourcesBuilder just keep a set of the paths it's already warned about?  That seems better than extending `DiagnosticsEngine`.

Reply:
> though it's usually somewhat of an antipattern for logic to ask the diagnostics engine whether it has emitted a particular diagnostic already — usually it would keep track of that itself.

I agree, if the internal storage of DiagnosticsEngine was an ordered set, this would be solved inherently without the calcite needing to do anything. that said, if there are reasons to keep it an array, then scanning it the way this PR does seems like it should change

Reply:
We could definitely change the internal storage for DiagnosticsEngine.  My only point was that it might be a surprising API if, for example, a client emitted two warnings but then there ended up being only one warning in the DiagnosticsEngine.  But I could see the coalescing being an option on DiagnosticsEngine, so that it was opt-in behavior — it's not a bad idea.  So my comment was more about what kind of behavior I think would be expected vs surprising in terms of the API.

If clients end up having to do a lot of work just to coalesce, then that seems like a strong argument for making it at least an option in DiagnosticsEngine.